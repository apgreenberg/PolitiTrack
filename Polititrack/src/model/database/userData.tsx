import { AllUserData } from './userInterfaces';

/*
The internal database
This will begin empty and the user will add data to it as they use the app
It will contain the same information as the external database, using AsyncStorage
Since AsyncStorage only stores strings, we must also store a more the data in a more readily available format
*/
const InternalDatabase: AllUserData = {
  users: [],
};

export default InternalDatabase;
