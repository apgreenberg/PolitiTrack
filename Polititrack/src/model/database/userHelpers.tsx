import AsyncStorage from '@react-native-async-storage/async-storage';

import { UserInfo } from './userInterfaces';
import InternalDatabase from './userData';

/**
 * Updates the internal database stored in userData.tsx with any recent changes
 */
export const updateInternalDatabase = async (): Promise<void> => {
  let allUsersString: string | null = null;
  try {
    allUsersString = await AsyncStorage.getItem('allUsers');
  } catch (e) {
    console.log(e);
  }
  if (allUsersString !== null) {
    let allUsers: UserInfo[] = JSON.parse(allUsersString).users;
    InternalDatabase.users = allUsers;
  }
};

/*
Updates the external database stored in AsyncStorage with any recent changes
*/
export const updateExternalDatabase = async (): Promise<void> => {
  try {
    await AsyncStorage.setItem('allUsers', JSON.stringify(InternalDatabase));
  } catch (e) {
    console.log(e);
  }
};
