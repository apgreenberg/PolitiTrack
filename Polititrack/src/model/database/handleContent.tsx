import { updateExternalDatabase } from './userHelpers';
import { OnboardContent, UserContent, UserInfo } from './userInterfaces';

/*
In the future, this will be generalized to work for bills and legislators as well
*/
/**
 * Adds the requested content to the user
 * @param userInfo - the user to add the content to
 *  * @param type - bills, billTypes, legislatures, or legislators
 * @param toAdd - the content to add
 */
export const addContentToDatabase = async (
  userInfo: UserInfo,
  type: keyof UserContent,
  toAdd: OnboardContent
): Promise<void> => {
  if (!alreadyContains(userInfo, type, toAdd)) {
    userInfo.data[type].push(toAdd);
    alert('Successfully added ' + toAdd.name);
  } else {
    alert('You have already added ' + toAdd.name);
  }
  await updateExternalDatabase();
};

/**
 * Removes the requested content from the user
 * @param userInfo  - the user to remove the content from
 * @param type - bills, billTypes, legislatures, or legislators
 * @param toRemoveId - the id of the content in the users list of content
 */
export const removeContentFromDatabase = async (
  userInfo: UserInfo,
  type: keyof UserContent,
  toRemoveId: string
): Promise<void> => {
  for (let i = 0; i < userInfo.data[type].length; i++) {
    if (userInfo.data[type][i].id === toRemoveId) {
      userInfo.data[type].splice(i, 1);
      break;
    }
  }
  await updateExternalDatabase();
};

/**
 * Helper function - checks if a certain content element already exists in the user's data
 * @param userInfo - the current user's info
 * @param type - the type of content to check
 * @param toAdd - the item the user is attempting to add
 * @returns - true if toAdd is already in the user's data, false otherwise
 */
export const alreadyContains = (
  userInfo: UserInfo,
  type: keyof UserContent,
  toAdd: OnboardContent
): boolean => {
  for (let content of userInfo.data[type]) {
    if (content.id === toAdd.id) {
      return true;
    }
  }
  return false;
};

/**
 * Updates the "lastUpdated" field in the content item
 * @param content - the content item to update
 */
export const updateLastUpdated = async (content: OnboardContent) => {
  content.lastUpdated = new Date().getTime();
  await updateExternalDatabase();
};
