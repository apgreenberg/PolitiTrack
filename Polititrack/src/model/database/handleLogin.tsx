import AsyncStorage from '@react-native-async-storage/async-storage';

import InternalDatabase from './userData';
import { updateExternalDatabase } from './userHelpers';
import { UserInfo } from './userInterfaces';

/**
 * Attempts to find a user in the internal database
 * @param email - the email of the user to find
 * @param password - the password of the user to find
 * @returns - the user if they are found, null otherwise
 */
export const findUser = (email: string, password: string): UserInfo | null => {
  const allUsers = InternalDatabase.users;
  for (let user of allUsers) {
    if (user.email === email && user.password === password) {
      return user;
    }
  }
  return null;
};

/**
 * Attempts to sign in the current user by updating the
 * 'currentUserInfo' record in AsyncStorage to the user's info
 * @param foundUserInfo - the info of the user to sign in
 */
export const attemptSignIn = async (foundUserInfo: UserInfo): Promise<void> => {
  try {
    await AsyncStorage.setItem(
      'currentUserInfo',
      JSON.stringify(foundUserInfo)
    );
  } catch (e) {
    console.log(e);
  }
};

/**
 * Attempts to sign out the current user by updating the
 * 'currentUserInfo' record in AsyncStorage to null
 */
export const attemptSignOut = async (): Promise<void> => {
  try {
    await AsyncStorage.removeItem('currentUserInfo');
  } catch (e) {
    console.log(e);
  }
};

/**
 * Attempts to delete a user's account
 * @param id - the id of the user's account to delete
 */
export const attemptDeleteAccount = async (id: number): Promise<void> => {
  for (let i: number = 0; i < InternalDatabase.users.length; i++) {
    if (InternalDatabase.users[i].id === id) {
      InternalDatabase.users.splice(i, 1);
    }
  }
  await updateExternalDatabase();
};

/*
Pulls the user info of the last signed in user from async storage, so user remains signed in
when reopening the app
*/
export const pullCurrentUserInfo = async (): Promise<UserInfo | null> => {
  let userInfo: UserInfo | null = null;
  let userInfoString: string | null = null;
  try {
    userInfoString = await AsyncStorage.getItem('currentUserInfo');
  } catch (e) {
    console.log(e);
  }
  if (userInfoString !== null) {
    userInfo = JSON.parse(userInfoString);
  }
  return userInfo;
};
