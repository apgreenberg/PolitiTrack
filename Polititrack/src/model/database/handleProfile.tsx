import { updateExternalDatabase } from './userHelpers';
import { UserInfo } from './userInterfaces';

export const updateProfile = (
  userInfo: UserInfo,
  type: keyof UserInfo,
  newValue: string
): void => {
  if (type === 'name' || type === 'email' || type === 'password') {
    userInfo[type] = newValue;
  }
  updateExternalDatabase();
};
