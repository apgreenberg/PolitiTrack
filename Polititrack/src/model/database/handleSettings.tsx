import {
  removeJurisdictionUpdateNotifications,
  removeBillUpdateNotifications,
  addJurisdictionUpdateNotifications,
  addBillUpdateNotifications,
} from '../background/notifications';
import { updateExternalDatabase } from './userHelpers';
import { UserInfo, UserSettings } from './userInterfaces';

/**
 * Handles updating the setting and the external database
 * @param userInfo - the current user's info
 * @param type - the type of setting to change
 * @param newValue - the new value of the setting
 */
export const updateSetting = async (
  userInfo: UserInfo,
  type: keyof UserSettings,
  newValue: boolean
): Promise<void> => {
  userInfo.settings[type] = newValue;
  await updateExternalDatabase();
};

/**
 * Handles toggling a setting
 * @param userInfo - the current user's info
 * @param type - the type of setting to change
 * @param callback - a callback function for updating the value state
 */
export const toggleSetting = async (
  userInfo: UserInfo,
  type: keyof UserSettings,
  callback: React.Dispatch<React.SetStateAction<boolean>>
): Promise<void> => {
  if (userInfo.settings[type]) {
    await updateSetting(userInfo, type, false);
    if (type === 'updateJurisdictions') {
      await removeJurisdictionUpdateNotifications();
    } else if (type === 'updateBills') {
      await removeBillUpdateNotifications();
    }
    callback(false);
  } else {
    await updateSetting(userInfo, type, true);
    if (type === 'updateJurisdictions') {
      await addJurisdictionUpdateNotifications();
    } else if (type === 'updateBills') {
      await addBillUpdateNotifications();
    }
    callback(true);
  }
};
