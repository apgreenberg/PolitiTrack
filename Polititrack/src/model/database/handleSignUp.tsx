import InternalDatabase from './userData';
import { UserInfo } from './userInterfaces';
import { updateExternalDatabase } from './userHelpers';

/**
 * Gets the next available user ID
 * @returns - the user ID, either 1 if they are the first user, or 1 + the user ID of the last user
 */
export const getNextId = (): number => {
  if (InternalDatabase.users.length === 0) {
    return 1;
  }
  return InternalDatabase.users[InternalDatabase.users.length - 1].id + 1;
};

/**
 * Adds a user to the internal and external databases
 * @param newUser - the user to add
 */
export const addUser = async (newUser: UserInfo): Promise<void> => {
  InternalDatabase.users.push(newUser);
  await updateExternalDatabase();
};
