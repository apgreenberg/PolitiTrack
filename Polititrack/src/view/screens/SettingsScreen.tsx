import React from 'react';
import { View, Text, Switch, SafeAreaView, StyleSheet } from 'react-native';
import 'react-native-gesture-handler';

import { selectCurrentUser, useAppSelector } from '../../redux/hooks';

import { toggleSetting } from '../../model/database/handleSettings';
import { UserInfo } from '../../model/database/userInterfaces';

import CustomButton from '../components/CustomButton';

import text from '../styles/text';
import colors from '../styles/colors';

/**
 * Allows the user to track updates to their saves jurisdictions/bills
 * This page's functionality is a week 4 objective
 * @returns - the functional components which represent the screen
 */
const SettingsScreen: React.FC = () => {
  const currentUser: UserInfo = useAppSelector(selectCurrentUser);
  const [jurisdictionSetting, setJurisdictionSetting] = React.useState(
    currentUser.settings.updateJurisdictions
  );
  const [billSetting, setBillSetting] = React.useState(
    currentUser.settings.updateBills
  );

  return (
    <SafeAreaView>
      <View style={{ paddingLeft: 10, paddingRight: 10 }}>
        <View
          style={{ paddingTop: 10, paddingBottom: 10, borderBottomWidth: 3 }}
        >
          <Text style={[text.largeText, { textAlign: 'left' }]}>
            Notifications
          </Text>
        </View>
        <View style={styles.settingsRow}>
          <View style={{ width: '85%', left: 15, paddingRight: 10 }}>
            <Text style={[text.mediumText, { textAlign: 'left' }]}>
              Notify me when one of my saved jurisdictions becomes active
            </Text>
          </View>
          <View style={{ right: 0 }}>
            <Switch
              testID={'jurisdictionSetting'}
              trackColor={{
                true: colors.highlightedSecondaryColor,
                false: colors.secondaryColor,
              }}
              thumbColor={jurisdictionSetting ? 'white' : colors.primaryColor}
              ios_backgroundColor={colors.secondaryColor}
              onValueChange={() =>
                toggleSetting(
                  currentUser,
                  'updateJurisdictions',
                  setJurisdictionSetting
                )
              }
              value={jurisdictionSetting}
            />
          </View>
        </View>
        <View style={styles.settingsRow}>
          <View style={{ width: '85%', left: 15, paddingRight: 10 }}>
            <Text style={[text.mediumText, { textAlign: 'left' }]}>
              Notify me when one of my saved bills is updated
            </Text>
          </View>
          <View style={{ right: 0 }}>
            <Switch
              testID={'billSetting'}
              trackColor={{
                true: colors.highlightedSecondaryColor,
                false: colors.secondaryColor,
              }}
              thumbColor={billSetting ? 'white' : colors.primaryColor}
              ios_backgroundColor={colors.secondaryColor}
              onValueChange={() =>
                toggleSetting(currentUser, 'updateBills', setBillSetting)
              }
              value={billSetting}
            />
          </View>
        </View>
        <View style={{ paddingTop: 20, height: '40%' }}>
          <CustomButton type="Update Account Information" />
        </View>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  settingsRow: {
    width: '100%',
    flexDirection: 'row',
    paddingTop: 5,
    paddingBottom: 5,
    borderBottomWidth: 1,
    borderBottomColor: colors.secondaryColor,
  },
});
export default SettingsScreen;
