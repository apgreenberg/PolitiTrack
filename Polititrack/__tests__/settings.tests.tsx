import React from 'react';
import TestRenderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';
import { MockedProvider } from '@apollo/client/testing';
import SettingsScreen from '../src/view/screens/SettingsScreen';
import { useSelectorMock, userData } from './common';

import * as handleSettings from '../src/model/database/handleSettings';
let mockToggleSetting = jest
  .spyOn(handleSettings, 'toggleSetting')
  .mockImplementation(() => Promise.resolve());

/**
 * Tests the settings page
 */
describe('Tests the settings screen and toggling the notifications', () => {
  useSelectorMock.mockReturnValue(userData);
  beforeEach(() => {
    mockToggleSetting.mockClear();
  });
  test('Viewing the settings page', () => {
    const { toJSON } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <SettingsScreen />
      </MockedProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });
  test('Toggle the bill notifications', () => {
    const { getByTestId } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <SettingsScreen />
      </MockedProvider>
    );
    fireEvent(getByTestId('jurisdictionSetting'), 'onValueChange');
    expect(mockToggleSetting).toHaveBeenCalled();
  });
  test('Toggle the jurisdiction notifications', () => {
    const { getByTestId } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <SettingsScreen />
      </MockedProvider>
    );
    fireEvent(getByTestId('billSetting'), 'onValueChange');
    expect(mockToggleSetting).toHaveBeenCalled();
  });
});
