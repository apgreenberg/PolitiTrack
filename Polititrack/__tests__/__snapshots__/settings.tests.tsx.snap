// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests the settings screen and toggling the notifications Viewing the settings page 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <View
      style={
        Object {
          "borderBottomWidth": 3,
          "paddingBottom": 10,
          "paddingTop": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 24,
              "textAlign": "center",
            },
            Object {
              "textAlign": "left",
            },
          ]
        }
      >
        Notifications
      </Text>
    </View>
    <View
      style={
        Object {
          "borderBottomColor": "#9ca7b8",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "paddingBottom": 5,
          "paddingTop": 5,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "left": 15,
            "paddingRight": 10,
            "width": "85%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "textAlign": "center",
              },
              Object {
                "textAlign": "left",
              },
            ]
          }
        >
          Notify me when one of my saved jurisdictions becomes active
        </Text>
      </View>
      <View
        style={
          Object {
            "right": 0,
          }
        }
      >
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#216ba3"
          style={
            Array [
              Object {
                "height": 31,
                "width": 51,
              },
              Object {
                "backgroundColor": "#9ca7b8",
                "borderRadius": 16,
              },
            ]
          }
          testID="jurisdictionSetting"
          thumbTintColor="white"
          tintColor="#9ca7b8"
          value={true}
        />
      </View>
    </View>
    <View
      style={
        Object {
          "borderBottomColor": "#9ca7b8",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "paddingBottom": 5,
          "paddingTop": 5,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "left": 15,
            "paddingRight": 10,
            "width": "85%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 16,
                "textAlign": "center",
              },
              Object {
                "textAlign": "left",
              },
            ]
          }
        >
          Notify me when one of my saved bills is updated
        </Text>
      </View>
      <View
        style={
          Object {
            "right": 0,
          }
        }
      >
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#216ba3"
          style={
            Array [
              Object {
                "height": 31,
                "width": 51,
              },
              Object {
                "backgroundColor": "#9ca7b8",
                "borderRadius": 16,
              },
            ]
          }
          testID="billSetting"
          thumbTintColor="white"
          tintColor="#9ca7b8"
          value={true}
        />
      </View>
    </View>
    <View
      style={
        Object {
          "height": "40%",
          "paddingTop": 20,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#dae8eb",
            "borderBottomWidth": 2,
            "borderColor": "#9ca7b8",
            "borderLeftWidth": 1,
            "borderRadius": 5,
            "borderRightWidth": 1,
            "borderTopWidth": 2,
            "height": "90%",
            "justifyContent": "center",
            "opacity": 1,
            "width": "90%",
          }
        }
        testID="customButton"
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "textAlign": "center",
            }
          }
        >
          Update Account Information
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
